packages/wrangler/src/pipelines/client.ts

```diff
@@ -96,44 +102,84 @@ export type PermissionGroup = {
 	scopes: string[];
 };
 
+interface S3AccessKey {
+	accessKeyId: string;
+	secretAccessKey: string;
+}
+
 // Generate a Service Token to write to R2 for a pipeline
 export async function generateR2ServiceToken(
-	label: string,
 	accountId: string,
-	bucket: string
-): Promise<ServiceToken> {
-	const res = await fetchResult<PermissionGroup[]>(
-		`/user/tokens/permission_groups`,
-		{
-			method: "GET",
-		}
-	);
-	const perm = res.find(
-		(g) => g.name == "Workers R2 Storage Bucket Item Write"
-	);
-	if (!perm) {
-		throw new Error("Missing R2 Permissions");
-	}
-
-	// generate specific bucket write token for pipeline
-	const body = JSON.stringify({
-		policies: [
-			{
-				effect: "allow",
-				permission_groups: [{ id: perm.id }],
-				resources: {
-					[`com.cloudflare.edge.r2.bucket.${accountId}_default_${bucket}`]: "*",
-				},
-			},
-		],
-		name: label,
+	bucketName: string,
+	pipelineName: string
+): Promise<S3AccessKey> {
+	let server: http.Server;
+	let loginTimeoutHandle: ReturnType<typeof setTimeout>;
+	const timerPromise = new Promise<S3AccessKey>((_, reject) => {
+		loginTimeoutHandle = setTimeout(() => {
+			server.close();
+			clearTimeout(loginTimeoutHandle);
+			reject(
+				new UserError(
+					"Timed out waiting for authorization code, please try again."
+				)
+			);
+		}, 120000); // wait for 120 seconds for the user to authorize
 	});
 
-	return await fetchResult<ServiceToken>(`/user/tokens`, {
-		method: "POST",
-		headers: API_HEADERS,
-		body,
+	const loginPromise = new Promise<S3AccessKey>((resolve, reject) => {
+		server = http.createServer(async (request, response) => {
+			assert(request.url, "This request doesn't have a URL"); // This should never happen
+
+			if (request.method !== "GET") {
+				response.writeHead(405);
+				response.end("Method not allowed.");
+				return;
+			}
+
+			const { pathname, searchParams } = new URL(
+				request.url,
+				`http://${request.headers.host}`
+			);
+
+			if (pathname !== "/") {
+				response.writeHead(404);
+				response.end("Not found.");
+				return;
+			}
+
+			// Retrieve values from the URL parameters
+			const accessKeyId = searchParams.get("access-key-id");
+			const secretAccessKey = searchParams.get("secret-access-key");
+
+			if (!accessKeyId || !secretAccessKey) {
+				reject(new UserError("Missing required URL parameters"));
+				return;
+			}
+
+			resolve({ accessKeyId, secretAccessKey } as S3AccessKey);
```

Comment:
Sure, I am trying to keep a lot of internal links out of the wrangler code base but if you think they would be a good addition here I can add some. 

In short, we can't use the standard OAuth flow for OAuth because those tokens are both short-lived and unable to communicate with R2 via the S3 API. These are both requirements that need to be met since these tokens are used by Pipelines service to write files into a customer R2 bucket on their behalf. More details can be found here: https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/src/user/user.ts#L942-L1039

Reply:
That would be great to add to the docblock at the top of the function. I'm just trying to avoid a situation where someone new to this feature comes across this code and has no idea why it's different from our OAuth flow. I think we can give a decent explanation without resorting to internal links.

Reply:
Added a docblock describing what it does and why we do it this way.